### ISSUE TITLE
TypeScript support for type-checking

### ISSUE THREAD

**Issue by jeromegn** [swc-project/swc#126]
I've successfully used swc to compile typescript files, that's great. I'm thoroughly impressed!

However, I noticed it compiled "wrong" typescript. Example:

```typescript
function foo(): string {
}
```

 ^ is invalid typescript (the function actually returns `void`), but this will still compile fine with swc.

I'm assuming this is a lot more complicated to make happen. You'd have to check keep track of the type signature of every function and variables to compare and check type validity.

**Comment by kdy1** [comment]
It was intentional as babel does not do type checking, but I'm considering implementing it. Once implemented, swc can replace tsc.

**Comment by sledorze** [comment]
@kdy1 have you evaluated how much a moving target typescript really is?

**Comment by kdy1** [comment]
@sledorze I didn't, but in [one blog post](https://blogs.msdn.microsoft.com/typescript/2018/08/27/typescript-and-babel-7/) they said they co-operated with babel team to make babel understand typescript. So I think trancompiler is not that hard to follow. 

I'm not sure about type system though. Type system of typescript seems to be changing frequently..

**Comment by sledorze** [comment]
@kdy1 yes the type system is changing frequently and is not sound, full of corner cases; I'm guessing that it's a herculean job..

**Comment by dgoldstein0** [comment]
Typescript's type system gets new improvements with every minor version release.  I don't see how it'd be a viable strategy for an alternative implementation to keep up with them - even if you have enough resources to keep up, you'd have to match it quirk for quirk to actually be compatible.

A lot of people separate transpiling and typechecking in their tools anyway - with some webpack plugins (awesome-typescript-loader) even forking separate processes to report typecheck errors outside the normal rebuild flow.  So I don't think this is a necessary feature, though certainly a project that just tried to speed up typescript's typechecking would be appreciated.

**Comment by jeromegn** [comment]
That makes sense! I didn't realize it changed so often. Now I just wish typescript would want to contribute and use this instead of their own checker.

Feel free to disregard this issue :)
